dips/dip-160.md

```diff
@@ -0,0 +1,59 @@
+---
+dip: 160
+title: Off-chain API ping command
+authors: Xiao Li (@xli)
+status: Draft
+discussions-to: https://github.com/diem/dip/issues/160
+type: Informational
+created: 04/08/2021
+requires: DIP-1
+---
+
+import useBaseUrl from '@docusaurus/useBaseUrl';
+
+# Summary
+
+This DIP describes an off-chain API v2 ping command that makes it easier for clients to detect network and protocol problems.
+
+# Abstract
+
+To improve the ability of off-chain service handling network problems and protocol errors, we’d like to introduce a ping command for client to:
+1. detect dead connections to a remote off-chain service.
+2. measure the latency to a remote off-chain service.
+3. detect off-chain protocol errors to a remote off-chain service, for example: when a remote off-chain service rotates their compliance key, and client used old compliance public key to verify the response message, the ping command can be used to debug the error and verify the fix.
+
+
+# Specification
+
+The PingCommand object definition:
+
+| Field       | Type | Required? | Description                  |
+|-------------|------|-----------|------------------------------|
+| _ObjectType | str  | Y         | The fixed string PingCommand |
+
+
+An example `CommandRequestObject` JSON message with `PingCommand`:
+
+```
+{
+    "_ObjectType": "CommandRequestObject",
+    "command_type": "PingCommand",
+    "command": {
+        "_ObjectType": "PingCommand",
+    },
+    "cid": "12ce83f6-6d18-0d6e-08b6-c00fdbbf085a",
+}
+```
+
+And the response of the `PingCommand`:
+```
+{
+    "_ObjectType": "CommandResponseObject",
+    “status”: “success”,
+    "cid": "12ce83f6-6d18-0d6e-08b6-c00fdbbf085a"
+}
+```
+
+Note we don’t need a random message body in PingCommand for the server to response, because the off-chain API protocol already have a `cid` field defined in the `CommandRequestObject` and `CommandResponseObject` to serve the purpose of distinguish different command instance.
```

Comment:
Note, PingCommand does not require a random payload, as the Off-Chain Protocol already has a `cid` to uniquely distinguish different `CommandRequestObject`s and `CommandResponseObject`s.